{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Importing packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "import os, sys, csv\n",
    "from datetime import datetime\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from collections import OrderedDict, defaultdict, Counter\n",
    "import shutil, argparse, re\n",
    "from heapq import nlargest\n",
    "from operator import itemgetter\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import operator, statistics\n",
    "from string import ascii_uppercase\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.interpolate import interp1d\n",
    "\n",
    "TOPN = 10\n",
    "path = \"/Users/kyle/Desktop/\"\n",
    "NEW = '\\n'\n",
    "TAB = '\\t'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parsing command line arguments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = argparse.ArgumentParser()\n",
    "parser.add_argument('-i', '--input', nargs='+', help='name of sequencing reads file', required=True)\n",
    "parser.add_argument('-o', '--output',  help='name and pathway to output file', default=r'dFASTR_output ' + datetime.now().strftime(\" %m-%d-%Y\") + '.csv')\n",
    "parser.add_argument('-ha', '--histall', help='output all (both new and old) histograms')\n",
    "parser.add_argument('-nh', '--new_hist', help='output new histograms')\n",
    "parser.add_argument('-oh', '--old_hist', help='output old histograms')\n",
    "\n",
    "args = parser.parse_args()\n",
    "\n",
    "filenamelist = args.input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Vividict(dict):\n",
    "    def __missing__(self, key):\n",
    "        value = self[key] = type(self)()\n",
    "        return value\n",
    "\n",
    "def getloci(path,flist):\n",
    "    \"\"\"Find the loci from the headers of all FASTA files, return as list.\"\"\"\n",
    "    llist = []\n",
    "    for f in flist:  #in args.input:\n",
    "        for read in SeqIO.parse(path+f, 'fasta'):\n",
    "            llist.append(re.split('\\W',read.id)[0])\n",
    "    return list(set(llist))\n",
    "\n",
    "def findlocus(header,l):\n",
    "    \"\"\"Find the locus name in the header of a FASTA record, return as string.\"\"\"\n",
    "    templist = [x for x in l if x in header]\n",
    "    return sorted(templist, key=len)[-1]\n",
    "\n",
    "def findsize(header):\n",
    "    \"\"\"Find the size (=frequency) of sequence in the header of a FASTA record, return as integer.\"\"\"\n",
    "    sizelong = [word for word in header.split(';') if any(x in word for x in ['size','freq'])][0]\n",
    "    return int(sizelong.split('=')[-1])\n",
    "\n",
    "def topNsequences(d, N=12):\n",
    "    \"\"\"Find the top N most frequent sequences from a dictionary of all sequences, return as dictionary.\"\"\"\n",
    "    topNtuples = sorted(d.items(), key=itemgetter(1), reverse=True)[:N]\n",
    "    return dict(topNtuples)\n",
    "\n",
    "def lengthsofseqalleles(s):\n",
    "    \"\"\"Find the lengths of alleles from list of sequences s, return as list.\"\"\"\n",
    "    return [len(l) for l in s]\n",
    "\n",
    "def rulebasedcall_1(d, r=2): \n",
    "    \"\"\"Use a rule-based logic to call alleles.\n",
    "    d=dictionary of TOPN sequences: d[seq]=size\n",
    "    \"\"\"\n",
    "    c = list() # c = list of candidate alleles\n",
    "    allele_1 = max(d, key=lambda key: d[key])\n",
    "    topfrequency = d.pop(allele_1)\n",
    "    if topfrequency >= 10:\n",
    "        c.append(allele_1)\n",
    "        for read in d:\n",
    "            if d[read] >= 10 and len(read) > 100:\n",
    "                #if (len(allele_1) -  len(read)) % r == 0:\n",
    "                if len(read) >= len(allele_1):\n",
    "                    if d[read] >= (topfrequency * 0.3):\n",
    "                        c.append(read)\n",
    "                if len(read) < len(allele_1):\n",
    "                    if d[read] >= (topfrequency * 0.3):\n",
    "                        c.append(read)\n",
    "    if len(c)==1:\n",
    "        c = c*2\n",
    "    return c\n",
    "\n",
    "def lengths2consider(d):\n",
    "    \"\"\"Find the lengths (alleles) to consider from a dictionary of top N sequences.\n",
    "    d=dictionary of TOPN sequences: d[seq]=size\n",
    "    RETURNS: a list of the lengths (allels), excluding any outliers\n",
    "    \"\"\"\n",
    "    lengthlist = list(set([len(seq) for seq in d]))\n",
    "    median = statistics.median(lengthlist) # m = median (about)\n",
    "    return sorted([ x for x in lengthlist if 0.5*median <= x <= 2*median ])\n",
    "     \n",
    "def preplengthdictionary(flist, llist):\n",
    "    \"\"\"Populate a defaultdict to be used in collecting length-to-frequency data.\n",
    "    RETURNS: dictionary with intended use: length2freq[locus][filename][len] += int \"\"\"\n",
    "    l2f = Vividict()\n",
    "    for f in flist:  #in args.input:\n",
    "        for locus in llist:\n",
    "            l2f[locus][f] = defaultdict(int)\n",
    "    return l2f\n",
    "\n",
    "def collectData(path, flist, l2f, llist):\n",
    "    \"\"\"Collect sequence, size(frequency), length, and locus data from FASTA files in list of filenames.\n",
    "    RETURNS: dictionaries with intended use: length2freq[locus][filename][len] = freq   where freq is integer\n",
    "    RETURNS: dictionaries with intended use: seq2freq[locus][filename][seq] = freq \"\"\"\n",
    "    s2f = Vividict()\n",
    "    for f in flist:  #in args.input:\n",
    "        for read in SeqIO.parse(path+f, 'fasta'):\n",
    "            locus = findlocus(read.id,llist)\n",
    "            size = findsize(read.id)\n",
    "            length = len(read.seq)\n",
    "            seq = str(read.seq)\n",
    "            if size > 1:\n",
    "                s2f[locus][f][seq] = size\n",
    "                l2f[locus][f][length] += size\n",
    "    return s2f, l2f\n",
    "\n",
    "def collectFrequencyArrays(path, flist, llist, MAXLENGTH=1000):\n",
    "    \"\"\"Collect sequence, size(frequency), length, and locus data from FASTA files in list of filenames.\n",
    "    RETURNS: dictionaries with intended use: length2freq[locus][filename][len] = freq   where freq is integer\n",
    "    RETURNS: dictionaries with intended use: seq2freq[locus][filename][seq] = freq \"\"\"\n",
    "    lengtharray = Vividict()\n",
    "    for f in flist:  #in args.input:\n",
    "        for locus in llist:\n",
    "            lengtharray[locus][f] = np.zeros(MAXLENGTH, dtype=int)\n",
    "        for read in SeqIO.parse(path+f, 'fasta'):\n",
    "            locus = findlocus(read.id,llist)\n",
    "            size = findsize(read.id)\n",
    "            length = len(read.seq)\n",
    "            seq = str(read.seq)\n",
    "            if size > 1:\n",
    "                lengtharray[locus][f][length] += size\n",
    "    return lengtharray\n",
    "\n",
    "######## This is where the program starts ###############           \n",
    "\n",
    "#1. Make list of loci in file\n",
    "print(\"Finding loci...\", end=\"\")\n",
    "loci_list = getloci(path,filenamelist)\n",
    "\n",
    "#2a. Prep frequency arrays\n",
    "print(\"Collecting length frequencies...\", end=\"\")\n",
    "lengtharray = collectFrequencyArrays(path, filenamelist, loci_list)\n",
    "\n",
    "def displayhistograms(lengtharray, MINREADS=20):\n",
    "\n",
    "    \"\"\"Plot histograms of read length frequencies.\n",
    "    INPUT: lengtharray = dictionary that maps locus and filename to an array of read frequencies:\n",
    "            lengtharray[locus][filename] = array([0,0,0,0,...0,0])\n",
    "    RETURNS: dictionaries with intended use: seq2freq[locus][filename][seq] = freq \"\"\"\n",
    "\n",
    "    for locus in lengtharray: \n",
    "        plotn = 1\n",
    "        plt.figure()\n",
    "        for filename in lengtharray[locus]:\n",
    "            plt.title(\"Heterozygous\", fontsize=20, fontweight='bold')\n",
    "            plt.xlabel(\"Read/allele length (bp)\", fontsize=20, fontweight='bold')\n",
    "            plt.ylabel(\"Read frequency\", fontsize=20, fontweight='bold')\n",
    "            idx = np.nonzero(lengtharray[locus][filename]>=MINREADS) # idx = indices of read frequencies >= MINREADS\n",
    "            ax = plt.bar(idx[0],lengtharray[locus][filename][idx])\n",
    "            plt.xticks(idx[0], rotation=30, fontsize=13, fontweight='bold')\n",
    "            plt.yticks(fontweight='bold', fontsize=13)\n",
    "            plotn +=1\n",
    "        plt.show()\n",
    "        # plt.savefig('foo.png')\n",
    "# displayhistograms(lengtharray)\n",
    "\n",
    "#2. Prep length2freq dict\n",
    "print(\"prepping length2freq dictionary...\", end=\"\")\n",
    "length2freq = preplengthdictionary(filenamelist,loci_list) #intended use: length2freq[locus][filename][len] += int\n",
    "\n",
    "#3. Populate master data dictionaries\n",
    "print(\"collecting data...\", end=\"\")\n",
    "seq2freq, length2freq = collectData(path, filenamelist, length2freq, loci_list) \n",
    "\n",
    "def printoldschoolhistograms(lengtharray, seq2freq, MINREADS=20):\n",
    "    \"\"\"Plot histograms of read length frequencies.\n",
    "    INPUT: lengtharray = dictionary that maps locus and filename to an array of read frequencies:\n",
    "            lengtharray[locus][filename] = array([0,0,0,0,...0,0])\n",
    "    RETURNS: dictionaries with intended use: seq2freq[locus][filename][seq] = freq \"\"\"\n",
    "    for locus in lengtharray:  #in args.input:\n",
    "        print(\".\", end=\"\")\n",
    "        for filename in lengtharray[locus]:\n",
    "            with open(path+filename+locus+'histograms.txt','w') as f:\n",
    "                f.write('-'*100+NEW)\n",
    "                f.write('$locus = ' + locus + NEW)\n",
    "            # for filename in lengtharray[locus]:\n",
    "                print(\".\", end=\"\")\n",
    "                f.write('~'*100+NEW)\n",
    "                f.write('~'*100+NEW)\n",
    "                f.write('#sample = ' + filename + NEW)\n",
    "                idx = np.nonzero(lengtharray[locus][filename]>=MINREADS) # idx = indices of read frequencies >= MINREADS\n",
    "                maxreads = max(lengtharray[locus][filename])\n",
    "                for i in range(np.amin(idx)-10,np.amax(idx)+10):\n",
    "                    nstars = int((lengtharray[locus][filename][i]*100)/maxreads)\n",
    "                    f.write(\" \".join([locus, filename, str(i).rjust(3), '*'*nstars])+NEW)\n",
    "                    topNdict = topNsequences(seq2freq[locus][filename])\n",
    "                rank = 0\n",
    "                longestallele = max(lengthsofseqalleles(topNdict.keys()))\n",
    "                for record in topNdict:\n",
    "                    length = str(len(record)).ljust(3)\n",
    "                    freq = ('freq='+str(topNdict[record])).ljust(12)\n",
    "                    sequence = record.ljust(longestallele+3)\n",
    "                    srank = str(rank).rjust(2)\n",
    "                    f.write(\" \".join(['@',locus, filename, srank, freq, length, '()', sequence, length, srank, NEW]))\n",
    "                    rank += 1\n",
    "\n",
    "print(\"Writing histograms...\", end=\"\")\n",
    "\n",
    "printoldschoolhistograms(lengtharray, seq2freq)\n",
    "\n",
    "def namealleles(genotypedictionary):\n",
    "    allelenames = Vividict() # Use as allelenames[locus][seq]=allelename\n",
    "    for locus in genotypedictionary:\n",
    "        templist = [] # a temporary list of all alleles for a particular locus\n",
    "        for sample in genotypedictionary[locus]:\n",
    "            for allele in genotypedictionary[locus][sample]:\n",
    "                templist.append(allele)\n",
    "        templist = list(set(templist)) # uniquify list\n",
    "        templendict = defaultdict(list)\n",
    "        for seq in templist:\n",
    "            templendict[len(seq)].append(seq)\n",
    "        lcounter = 0\n",
    "        for length in templendict:\n",
    "            templendict[length].sort()\n",
    "            for seq in templendict[length]:\n",
    "                allelename = '_'.join([locus,str(length),ascii_uppercase[lcounter]])\n",
    "                allelenames[locus][seq] = allelename\n",
    "                lcounter += 1\n",
    "                print(locus, length, seq, allelenames[locus][seq])\n",
    "    return allelenames\n",
    "\n",
    "def outputallelenames(allelenamedict, path, alleleoutfilename): #output alleles to fasta file\n",
    "    seqlist = []\n",
    "    for locus in allelenames:\n",
    "        for alleleseq in allelenames[locus]:\n",
    "            header = allelenames[locus][alleleseq]\n",
    "            seqlist.append(SeqRecord(Seq(alleleseq),id=header, description=header))\n",
    "    with open(path+alleleoutfilename, 'w') as handle:\n",
    "        SeqIO.write(seqlist, handle, 'fasta')\n",
    "\n",
    "# allelenames = namealleles(oldschoolcalls)\n",
    "# outputallelenames(allelenames, path, 'allelenames.txt')\n",
    "\n",
    "#4. Call alleles          \n",
    "print(\"calling alleles...\", end=\"\")\n",
    "\n",
    "candidatealleles = Vividict() # intended use: length2freq[locus][filename][method] = [allele1,allele2]\n",
    "\n",
    "for locus in seq2freq:  #in args.input:\n",
    "    for filename in seq2freq[locus]:\n",
    "        topNdict = topNsequences(seq2freq[locus][filename])\n",
    "        topNlengths = lengths2consider(topNdict)\n",
    "\n",
    "        candidatealleles[locus][filename]['rulebasedcall_1'] = rulebasedcall_1(topNdict)\n",
    "\n",
    "def displaycalls(candidatealleles):\n",
    "    \"\"\"Plot histograms of read length frequencies.\n",
    "    INPUT: candidatealleles = dictionary that maps locus and filename to an array of read frequencies:\n",
    "            candidatealleles[locus][filename][callname] = [allele_1, allele_2]\"\"\"\n",
    "    print(\"\",)\n",
    "    print(\"{:<20} {:<20} {:<20}\".format('Filename','Locus','Alleles'))\n",
    "    for locus in candidatealleles:  #in args.input:\n",
    "        for filename in candidatealleles[locus]:\n",
    "            a = lengthsofseqalleles(candidatealleles[locus][filename]['Alleles'])         \n",
    "            print(\"{:<20} {:<20} {:<20}\".format(filename, locus, str(a)))\n",
    "\n",
    "# displaycalls(candidatealleles)\n",
    "\n",
    "print('done')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
